### PR TITLE
Resolved #234

### LINKED ISSUE #234: Back up an Azure VM using Azure Backup
##### SUMMARY

Currently, there is no module available for back up an Azure VM using Azure Backup 

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

azure_rm_backupAzureVM
azure_rm_backupAzureVM_info

##### ADDITIONAL INFORMATION

This module will introduce backup feature for azure vm and will be offering feature listed below
1. Enabling protection for the Azure VM
2. Trigger an on-demand backup for a protected Azure VM
3. Modify the backup configuration for a protected Azure VM
4. Stop protection but retain existing data
5. Stop protection and delete data

I am planning on writing the modules myself, but wanted to post an Issue ahead of time to solicit feedback and comments and to make sure no one else is already working on it

```yaml

```

### PR DESCRIPTION
**Pull request by suyeb786**
##### SUMMARY

This module will introduce a backup feature for azure VM and will be offering feature listed below

Enabling protection for the Azure VM
Trigger an on-demand backup for a protected Azure VM
Modify the backup configuration for a protected Azure VM
Stop protection but retain existing data
Stop protection and delete data

Fixes #234 
##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

azure_rm_backupazurevm
azure_rm_backupazurevm_info
##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by suyeb786** [comment]
Hi @Fred-sun,
My module is ready for Azure VM backup(#234).
Could you please review it.

File: plugins/modules/azure_rm_backupazurevm.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1', 'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = \
+    '''
+---
+module: azure_rm_backupazurevm
+version_added: '2.7.6'
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
version_added: '1.1.0'
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    - Stop protection and delete data.
+options:
+    resource_group:
+        description:
+            - The name of the resource group.
+        required: true
+        type: str
+    recovery_vault_name:
+        description:
+            - The name of the Azure Recovery Service Vault
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - The name of the Azure Recovery Service Vault。
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        required: true
+        type: str
+    resource_id:
+        description:
+            - Azure Virtual Machine Resource ID
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - Azure Virtual Machine Resource ID.
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        type: str
+    backup_policy_id:
+        description:
+            - Backup Policy ID present under Recovery Service Vault mentioned in recovery_vault_name field.
+        required: true
+        type: str
+    state:
+        description:
+            - Assert the state of the protection item.
+            - Use "create" for enabling protection for the Azure VM.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - Use C(create) for enabling protection for the Azure VM.
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+            - Use "update" for changing the policy of protection.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - Use C(update) for changing the policy of protection.
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+            - Use "stop" for stop protection but retain existing data.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - Use C(stop) for stop protection but retain existing data.
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+            - Use "delete" for stop protection and delete data.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - Use C(delete) for stop protection and delete data.
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            - Use 'backup' for on-demand backup.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]:
```suggestion
            - Use C(backup) for on-demand backup.
```

File: plugins/modules/azure_rm_backupazurevm_info.py [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1', 'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = \
+    '''
+---
+module: azure_rm_backupazurevm
+version_added: '2.7.6'
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm_info.py]:
```suggestion
version_added: '1.1.0'
```

File: plugins/modules/azure_rm_backupazurevm_info.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    - Get Recovery point details for protected items.
+options:
+    resource_group:
+        description:
+            - The name of the resource group.
+        required: true
+        type: str
+    recovery_vault_name:
+        description:
+            - The name of the Azure Recovery Service Vault
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm_info.py]:
```suggestion
            - The name of the Azure Recovery Service Vault.
```

File: plugins/modules/azure_rm_backupazurevm_info.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        required: true
+        type: str
+    resource_id:
+        description:
+            - Azure Virtual Machine Resource ID
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm_info.py]:
```suggestion
            - Azure Virtual Machine Resource ID.
```

File: plugins/modules/azure_rm_backupazurevm.py (hunk 1 above)
Comment (suyeb786) [reply plugins/modules/azure_rm_backupazurevm.py]: The version is updated.

**Comment by suyeb786** [comment]
Hi @Fred-sun,
Thank you for reviewing the module
The code is adjusted with the latest comments.

**Comment by Fred-sun** [comment]
> Hi @Fred-sun,
> Thank you for reviewing the module
> The code is adjusted with the latest comments.

Hi @suyeb786 I'm testing, and will feedback to you! Thank you very much!

**Comment by Fred-sun** [comment]
Hi @suyeb786 How did you get the backup_policy_id?
```e.g.
Suggestion change:  
    1. Rename *.yaml to .yml
    2. Add the module name azure_rm_backupazurevm in ./pr-pipeline.yml
    3 Update main.yml' "backup_policy_id" to "{{ backup_policy_id }}"

File: pr-pipelines.yml [hunk 12]
```diff
@@ -71,6 +71,8 @@ parameters:
     - "azure_rm_virtualnetwork"
     - "azure_rm_virtualnetworkgateway"
     - "azure_rm_virtualnetworkeepring"
+    - "azure_rm_backupazurevm"
+    - "azure_rm_backupazurevm_info"
```
Comment (Fred-sun) [inline pr-pipelines.yml]:
```suggestion

```

**Comment by Fred-sun** [comment]
Update **tests/integration/targets/azure_rm_backupazurevm/meta/main.yaml ** to **tests/integration/targets/azure_rm_backupazurevm/meta/main.yml **

**Comment by suyeb786** [comment]
Hi @Fred-sun,

In the recovery service vault, we have provision for creating backup policies.
Also, there is a default policy that is available by default when we create a recovery service vault in azure.

    "policyId": "/subscriptions/<subscription_id>/resourceGroups/<resource_group_name>/providers/microsoft.recoveryservices/vaults/<vault_name>/backupPolicies/DefaultPolicy"


replace <subscription_id>  ->> with subsription id
replace <resource_group_name> ->> with resouce group name
replace <vault_name> >> with your recovery vault name

eg.

"/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/testVaultRG/providers/microsoft.recoveryservices/vaults/testVault/backupPolicies/DefaultPolicy"

File: plugins/modules/azure_rm_backupazurevm_info.py [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+author:
+    - Suyeb Ansari (@suyeb786)
+    - Pallavi Chaudhari (@PallaviC2510)
+
+'''
+
+EXAMPLES = \
+    '''
+    - name: Get Recovery Point Details
+      azure_rm_backupazurevm:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm_info.py]:
```suggestion
      azure_rm_backupazurevm_info:
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+        if self.state == 'create' or self.state == 'update':
+            response = self.enable_update_protection_for_azure_vm()
+        if self.state == 'delete':
+            response = self.stop_protection_and_delete_data()
+        if self.state == 'stop':
+            response = self.stop_protection_but_retain_existing_data()
+        if self.state == 'backup':
+            response = self.trigger_on_demand_backup()
+        self.results['response'] = response
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py:267]: It is recommended to make the following changes in lines 245~254
```e.g.
        changed = False
        if self.state == 'create' or self.state == 'update':
            changed = True
            response = self.enable_update_protection_for_azure_vm()
        if self.state == 'delete':
            changed = True
            response = self.stop_protection_and_delete_data()
        if self.state == 'stop':
            changed = True
            response = self.stop_protection_but_retain_existing_data()
        if self.state == 'backup':
            changed = True
            response = self.trigger_on_demand_backup()
        self.results['response'] = response
        self.results['changed'] = changed

File: tests/integration/targets/azure_rm_backupazurevm/aliases [hunk 15]
```diff
@@ -0,0 +1,3 @@
+cloud/azure
+shippable/azure/group12
+destructive
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_backupazurevm/aliases:3]: Please help to make changes as follows. Thank you!
```suggestion
destructive
disabled
```

File: plugins/modules/azure_rm_backupazurevm.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+        return '2019-05-13' if self.state == 'create' or self.state == 'update' or self.state == 'delete' or self.state == 'stop' else '2016-12-01'
+
+    def get_url(self):
+        if self.state == 'create' or self.state == 'update' or self.state == 'delete' or self.state == 'stop':
+            return '/subscriptions' + '/' + self.subscription_id \
+                   + '/resourceGroups' + '/' + self.resource_group + '/providers' \
+                   + '/Microsoft.RecoveryServices' + '/vaults' + '/' \
+                   + self.recovery_vault_name \
+                   + '/backupFabrics/Azure/protectionContainers/' \
+                   + 'iaasvmcontainer;iaasvmcontainerv2;' + self.resource_group \
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py:191]: The string here is weird, why do you do this?

**Comment by Fred-sun** [comment]
It is recommended to update tasks as follows:
```e.g.
- name: Fix resource prefix
  set_fact:
    resource_group: 'MyResourceGroup'
    recovery_vault_name: 'MyRecoveryVault'
    resource_id: "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/Microsoft.Compute/virtualMachines/MyVM"
    backup_policy_id: "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/microsoft.recoveryservices/vaults/testVault/backupPolicies/ProdPolicy"

- name: Enabling/Updating protection for the Azure VM
  azure_rm_backupazurevm:
    resource_group: "{{ resource_group }}"
    recovery_vault_name: "{{ recovery_vault_name }}"
    resource_id: "{{ resource_id }}"
    backup_policy_id: "{{ backup_policy_id }}"
    state: "create"
  register: output

- name: Assert that output has changed
  assert:
    that:
      - output.changed
      
- name: Trigger an on-demand backup for a protected Azure VM
  azure_rm_backupazurevm:
    resource_group: "{{ resource_group }}"
    recovery_vault_name: "{{ recovery_vault_name }}"
    resource_id: "{{ resource_id }}"
    backup_policy_id: "{{ backup_policy_id }}"
    state: "backup"
  register: output

- name: Assert that output has changed
  assert:
    that:
      - output.changed

- name: Stop protection but retain existing data
  azure_rm_backupazurevm:
    resource_group: "{{ resource_group }}"
    recovery_vault_name: "{{ recovery_vault_name }}"
    resource_id: "{{ resource_id }}"
    backup_policy_id: "{{ backup_policy_id }}"
    state: "stop"
  register: output

- name: Assert that output has changed
  assert:
    that:
      - output.changed
      
- name: Get backup azure vm info
  azure_rm_backupazurevm_info:
    resource_group: "{{ resource_group }}"
    recovery_vault_name: "{{ recovery_vault_name }}"
    resource_id: "{{ resource_id }}"
  register: output

- name: Assert that output has changed
  assert:
    that:
      - output.response.id != None
      - output.response.name != None
   
- name: Stop protection and delete data
  azure_rm_backupazurevm:
    resource_group: "{{ resource_group }}"
    recovery_vault_name: "{{ recovery_vault_name }}"
    resource_id: "{{ resource_id }}"
    backup_policy_id: "{{ backup_policy_id }}"
    state: "delete"
  register: output

- name: Assert that output has changed
  assert:
    that:
      - output.changed

File: plugins/modules/azure_rm_backupazurevm.py (hunk 16 above)
Comment (suyeb786) [reply plugins/modules/azure_rm_backupazurevm.py:191]: Below is the API call for azure vm backup

PUT https://management.azure.com/Subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/testVaultRG/providers/Microsoft.RecoveryServices/vaults/testVault/backupFabrics/Azure/protectionContainers/iaasvmcontainer;iaasvmcontainerv2;testRG;testVM/protectedItems/vm;iaasvmcontainerv2;testRG;testVM?api-version=2019-05-13


containerName = "iaasvmcontainer;"+{name}
protectedItemName = "vm;"+ {name}

eg.

containerName = "iaasvmcontainer;iaasvmcontainerv2;testRG;testVM"
protectedItemName = "vm;iaasvmcontainerv2;testRG;testVM

containrName and ProtectedItemName is combination of resouceGroup + VMName + Container string

Comment (Fred-sun) [reply plugins/modules/azure_rm_backupazurevm.py:191]: I see! Thank you very much！

File: plugins/modules/azure_rm_backupazurevm.py [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+      azure_rm_backupazurevm:
+        resource_group: 'myResourceGroup'
+        recovery_vault_name: 'testVault'
+        resource_id: '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/Microsoft.Compute/virtualMachines/testVM'
+        state: 'stop'
+    - name: Stop protection and delete data
+      azure_rm_backupazurevm:
+        resource_group: 'myResourceGroup'
+        recovery_vault_name: 'testVault'
+        resource_id: '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/Microsoft.Compute/virtualMachines/testVM'
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]: The sentence is too long, please shorten it.

File: plugins/modules/azure_rm_backupazurevm.py [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+'''
+
+EXAMPLES = \
+    '''
+    - name: Enabling/Updating protection for the Azure VM
+      azure_rm_backupazurevm:
+        resource_group: 'myResourceGroup'
+        recovery_vault_name: 'testVault'
+        resource_id: '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/Microsoft.Compute/virtualMachines/testVM'
+        backup_policy_id: '/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myResourceGroup/providers/microsoft.recoveryservices/vaults/testVault/backupPolicies/ProdPolicy'
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm.py]: The sentence is too long, please shorten it.

File: plugins/modules/azure_rm_backupazurevm_info.py [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1', 'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = \
+    '''
+---
+module: azure_rm_backupazurevm
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_backupazurevm_info.py]:
```suggestion
module: azure_rm_backupazurevm_info
```

File: plugins/modules/azure_rm_backupazurevm.py (hunk 18 above)
Comment (suyeb786) [reply plugins/modules/azure_rm_backupazurevm.py]: Adjusted string with column size 80 as per pep8 standards

File: plugins/modules/azure_rm_backupazurevm.py (hunk 17 above)
Comment (suyeb786) [reply plugins/modules/azure_rm_backupazurevm.py]: Adjusted string with column size 80 as per pep8 standards

File: plugins/modules/azure_rm_backupazurevm_info.py (hunk 19 above)
Comment (suyeb786) [reply plugins/modules/azure_rm_backupazurevm_info.py]: The code is adjusted.
I apologize for the minor mistakes.

File: tests/sanity/ignore-2.10.txt [hunk 20]
```diff
@@ -707,5 +707,14 @@ plugins/modules/azure_rm_adserviceprincipal.py validate-modules:required_if-unkn
 plugins/modules/azure_rm_adserviceprincipal_info.py validate-modules:nonexistent-parameter-documented
 plugins/modules/azure_rm_adserviceprincipal_info.py validate-modules:required_if-requirements-unknown
 plugins/modules/azure_rm_adserviceprincipal_info.py validate-modules:required_if-unknown-key
+plugins/modules/azure_rm_backupazurevm.py validate-modules:missing-suboption-docs
+plugins/modules/azure_rm_backupazurevm.py validate-modules:parameter-type-not-in-doc
+plugins/modules/azure_rm_backupazurevm.py validate-modules:undocumented-parameter
+plugins/modules/azure_rm_backupazurevm.py validate-modules:invalid-ansiblemodule-schema
+plugins/modules/azure_rm_backupazurevm.py validate-modules:parameter-list-no-elements
```
Comment (Fred-sun) [inline tests/sanity/ignore-2.10.txt]: Please delete 710,711,712, 713 and 714 lines. and synchronize the changes tests/sanity/ignore-2.11.txt and tests/sanity/ignore-2.9.txt

Comment (suyeb786) [reply tests/sanity/ignore-2.10.txt]: Removed lines 710,711,712, 713 and 714 from tests/sanity/ignore-2.10.txt 
Also, tests/sanity/ignore-2.11.txt and tests/sanity/ignore-2.9.txt is synchronized with tests/sanity/ignore-2.10.txt

File: tests/sanity/ignore-2.9.txt [hunk 21]
```diff
@@ -183,5 +183,9 @@ plugins/modules/azure_rm_adpassword_info.py validate-modules:nonexistent-paramet
 plugins/modules/azure_rm_adpassword_info.py validate-modules:import-before-documentation
 plugins/modules/azure_rm_adserviceprincipal.py validate-modules:nonexistent-parameter-documented
 plugins/modules/azure_rm_adserviceprincipal_info.py validate-modules:nonexistent-parameter-documented
+plugins/modules/azure_rm_backupazurevm.py validate-modules:required_if-requirements-unknown
+plugins/modules/azure_rm_backupazurevm.py validate-modules:required_if-unknown-key
+plugins/modules/azure_rm_backupazurevm_info.py validate-modules:required_if-requirements-unknown
+plugins/modules/azure_rm_backupazurevm_info.py validate-modules:required_if-unknown-key
```
Comment (Fred-sun) [inline tests/sanity/ignore-2.9.txt]: It's my mistake, Please delete 186,187,188 and 189 lines; Thank you very much!

File: tests/integration/targets/azure_rm_backupazurevm/tasks/main.yml [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    resource_id: "{{ resource_id }}"
+    backup_policy_id: "{{ backup_policy_id }}"
+    state: "delete"
+  register: output
+  
+- name: Assert that output has changed
+  assert:
+    that:
+      - output.changed
+
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_backupazurevm/tasks/main.yml]: Please delete the last extra blank line！

File: tests/sanity/ignore-2.9.txt (hunk 21 above)
Comment (suyeb786) [reply tests/sanity/ignore-2.9.txt]: 186-189 lines are removed from tests/sanity/ignore-2.9.txt

File: tests/integration/targets/azure_rm_backupazurevm/tasks/main.yml (hunk 22 above)
Comment (suyeb786) [reply tests/integration/targets/azure_rm_backupazurevm/tasks/main.yml]: blank is removed from tasks/main/yml

**Comment by Fred-sun** [comment]
@suyeb786 Thank you for your supportive response. Basically, it has been changed almost!

**Comment by suyeb786** [comment]
Hello @Fred-sun ,

Thank you for your help and quick reviews, with your help I understand all core requirements for the ansible azure modules and it will be very helpful for me for feature development.

**Pull request merged**